File: firebase/functions/lib/events/notifications/email_event_reminder.dart
import 'dart:async';

import '../../on_request_method.dart';
import 'event_emails.dart';
import '../../utils/firestore_utils.dart';
import 'package:data_models/cloud_functions/requests.dart';
import 'package:data_models/firestore/event.dart';
import 'package:data_models/firestore/community.dart';

/// This request method is called from our scheduled task queue.
///
/// It checks to see if this type of event email has been sent to this user
/// before and if not, sends the email.
class EmailEventReminder extends OnRequestMethod<EmailEventReminderRequest> {
  EventEmails eventEmailUtils;

  EmailEventReminder({EventEmails? eventEmailUtils})
      : eventEmailUtils = eventEmailUtils ?? EventEmails(),
        super(
          'EmailEventReminder',
          (jsonMap) => EmailEventReminderRequest.fromJson(jsonMap),
        );

  static final emailReminderOffset = <EventEmailType, Duration>{
    EventEmailType.oneDayReminder: const Duration(days: 1),
    EventEmailType.oneHourReminder: const Duration(hours: 1),
  };

  static const _reminderDurationBuffer = Duration(minutes: 30);

  @override
  Future<String> action(EmailEventReminderRequest request) async {
    await firestore.runTransaction((transaction) async {
      final eventPath =
          'community/${request.communityId}/templates/${request.templateId}/events/${request.eventId}';
      final event = await firestoreUtils.getFirestoreObject(
        transaction: transaction,
        path: eventPath,
        constructor: (map) => Event.fromJson(map),
      );

      print('Attempting reminder email for $event');
      print('Current time: ${DateTime.now()}');
      final reminderOffset = emailReminderOffset[request.eventEmailType];

      if (reminderOffset == null) return;

      final timeUntilEvent = event.scheduledTime?.difference(DateTime.now());
      if (timeUntilEvent == null ||
          timeUntilEvent > (reminderOffset + _reminderDurationBuffer) ||
          timeUntilEvent < (reminderOffset - _reminderDurationBuffer)) {
        print('Time until event: $timeUntilEvent');
        print(
          'Event is not within $_reminderDurationBuffer notification window of $reminderOffset.',
        );
        return;
      }

      final community = await firestoreUtils.getFirestoreObject(
        path: 'community/${request.communityId}',
        constructor: (map) => Community.fromJson(map),
      );

      final suppressEmail = !(event.eventSettings?.reminderEmails ??
          community.eventSettingsMigration.reminderEmails ??
          true);
      if (suppressEmail) return;

      final participantsSnapshot =
          await firestore.collection('$eventPath/event-participants').get();
      final participants = participantsSnapshot.documents
          .map((doc) => Participant.fromJson(doc.data.toMap() ?? {}))
          .where(
            (participant) => participant.status == ParticipantStatus.active,
          )
          .toList();

      await eventEmailUtils.sendEmailsToUsers(
        transaction: transaction,
        event: event,
        eventPath: eventPath,
        userIds: participants.map((participant) => participant.id).toList(),
        emailType: request.eventEmailType!,
      );
    });

    return '';
  }
}

File: firebase/functions/src/events/notifications/email_event_reminder.ts
import { OnRequestMethod } from '../../on_request_method';
import { firestore, firestoreUtils } from '../../utils/infra/firestore_utils';
import { Event, Community, Participant, ParticipantStatus, EventEmailType } from '../../types';

interface EmailEventReminderRequest {
  communityId: string;
  templateId: string;
  eventId: string;
  eventEmailType: EventEmailType;
}

const REMINDER_DURATION_BUFFER_MS = 30 * 60 * 1000; // 30 minutes

export class EmailEventReminder extends OnRequestMethod<EmailEventReminderRequest> {
  static readonly emailReminderOffsetMs: Record<string, number> = {
    [EventEmailType.oneDayReminder]: 24 * 60 * 60 * 1000,
    [EventEmailType.oneHourReminder]: 60 * 60 * 1000,
  };

  constructor() {
    super('EmailEventReminder', (jsonMap) => jsonMap as EmailEventReminderRequest);
  }

  async action(request: EmailEventReminderRequest): Promise<string> {
    await firestore.runTransaction(async (transaction) => {
      const eventPath = `community/${request.communityId}/templates/${request.templateId}/events/${request.eventId}`;
      const event: Event = await firestoreUtils.getFirestoreObject({
        transaction,
        path: eventPath,
        constructor: (map) => map as unknown as Event,
      });

      console.log('Attempting reminder email for event:', event.id);
      const reminderOffsetMs = EmailEventReminder.emailReminderOffsetMs[request.eventEmailType];
      if (reminderOffsetMs === undefined) return;

      const timeUntilEvent = event.scheduledTime
        ? event.scheduledTime.getTime() - Date.now()
        : null;
      if (
        timeUntilEvent === null ||
        timeUntilEvent > reminderOffsetMs + REMINDER_DURATION_BUFFER_MS ||
        timeUntilEvent < reminderOffsetMs - REMINDER_DURATION_BUFFER_MS
      ) {
        console.log('Event is not within notification window of reminder offset:', reminderOffsetMs);
        return;
      }

      const community: Community = await firestoreUtils.getFirestoreObject({
        path: `community/${request.communityId}`,
        constructor: (map) => map as unknown as Community,
      });

      const suppressEmail = !(
        event.eventSettings?.reminderEmails ??
        (community as any).eventSettingsMigration?.reminderEmails ??
        true
      );
      if (suppressEmail) return;

      const participantSnaps = await firestore
        .collection(`${eventPath}/event-participants`)
        .get();
      const participants = participantSnaps.docs
        .map((doc) => ({
          ...(firestoreUtils.fromFirestoreJson(doc.data()) as unknown as Participant),
          id: doc.id,
        }))
        .filter((p) => p.status === ParticipantStatus.active);

      // Lazy import to avoid circular dependency
      const { EventEmails } = await import('./event_emails');
      await new EventEmails().sendEmailsToUsers({
        event,
        eventPath,
        userIds: participants.map((p) => p.id!),
        emailType: request.eventEmailType,
      });
    });

    return '';
  }
}
